content/BCPatterns/generic-method-pattern/index.md

```diff
@@ -0,0 +1,327 @@
+
+This is the relevant code for the event layer:
+
+```AL
+codeunit 53100 "WLD BlockCustomer Meth"
+{
+    internal procedure BlockCustomer(var Cust: Record Customer; HideDialog: Boolean);
+    var
+        Handled: Boolean;
+    begin
+        ...
+        OnBeforeBlockCustomer(Cust, Handled);
+        ...
+        OnAfterBlockCustomer(Cust);
+        ...
+    end;
+...    
+    [IntegrationEvent(false, false)]
+    local procedure OnBeforeBlockCustomer(var Cust: Record Customer; var Handled: Boolean);
+    begin
+    end;
+
+    [IntegrationEvent(false, false)]
+    local procedure OnAfterBlockCustomer(var Cust: Record Customer);
+    begin
+    end;
+}
+```
+
+*The method layer*
+The last layer is obviously where the business logic will be written.
+
+The relevant part is:
+
+```AL
+codeunit 53100 "WLD BlockCustomer Meth"
+{
+    internal procedure BlockCustomer(var Cust: Record Customer; HideDialog: Boolean);
+    var
+        Handled: Boolean;
+    begin
+        ...
+        DoBlockCustomer(Cust, Handled);
+        ...
+    end;
+
+    local procedure DoBlockCustomer(var Cust: Record Customer; Handled: Boolean);
+    begin
+        if Handled then
+            exit;
+
+        Cust.Blocked := Cust.Blocked::All;
+        Cust.Modify(true);
+    end;
+...
+}
+```
+Usually indicated with a "do"-function, the business logic takes place in that procedure.  Obviously, when you have a decent amount of code, it's recommended that you make it readable by applying all the Best Practices in terms of readability in the codeunit.  Though, a few pointers here:
+- keep the cyclomatic complexity low
+  - one line (functioncall) after an IF-clause
+  - one line (functioncall) after a repeat
+- readable function calls
+- ...
+
+**Call out to the method from a table**
+
+Now, there is a reason that the one global procedure in the method-codeunit is `internal`.  We shouldn't be calling this procedure directly.  We should be calling it through it's "class" (usually, a table can be seen as a class).  That means, we would have to create a public (not internal) procedure on table-level, which can be used in the business logic.
+
+In our example above, this could be that table extension:
+
+```AL
+tableextension 53100 "Customer Ext BASE" extends Customer
+{
+    procedure BlockCustomer(HideDialog: Boolean)
+    var
+        WLDBlockCustomerMeth: Codeunit "WLD BlockCustomer Meth";
+    begin
+        WLDBlockCustomerMeth.BlockCustomer(Rec, HideDialog);
+    end;
+
+}
+```
+
+This practice improves readability.  In fact, by doing this, you just extended the suggesions-list in VSCode (IntelliSense) indicating a new method that your class can do.  This is very convenient for the developer that might need your new method.
+
+*Note - it could very well be that there simply isn't any table that can act as a class for our method.  In that case, you could use a codeunit as well.*
+
```

Comment:
Would it be an idea to point to the system app which uses the pattern of two codeunits: one codeunit with public functions and an internal codeunit for the implementation.

Reply:
That is a completely different pattern - which could also be described at some point.

As such, the implementation codeunit of that pattern would/could be multiple methods.  Only, they do it to "protect" the code, while in this pattern, we're exposing the business logic.

So - in my opinion, it should be described separately.

Reply:
Just reacting to the quote 'you could use a codeunit as well'. Which is indeed a different pattern. ☺️

Reply:
I know ;-).
I hope that pattern will be subscribed by Microsoft - and obviously at that point, we'll be able to refer to it as well..